frame/staking/src/voter_bags.rs

```diff
@@ -0,0 +1,897 @@
+		}
+		self.tail = Some(id.clone());
+
+		crate::VoterBagFor::<T>::insert(id, self.bag_upper);
+	}
+
+	/// Remove a voter node from this bag.
+	///
+	/// This is private on purpose because it doesn't check whether this bag contains the voter in
+	/// the first place. Generally, use [`VoterList::remove`] instead.
+	///
+	/// Storage note: this modifies storage, but only for adjacent nodes. You still need to call
+	/// `self.put()` and `node.put()` after use.
+	fn remove_node(&mut self, node: &Node<T>) {
+		node.excise();
+
+		// clear the bag head/tail pointers as necessary
+		if self.head.as_ref() == Some(&node.voter.id) {
+			self.head = node.next.clone();
+		}
+		if self.tail.as_ref() == Some(&node.voter.id) {
+			self.tail = node.prev.clone();
+		}
+	}
+}
+
+/// A Node is the fundamental element comprising the doubly-linked lists which for each bag.
+#[derive(Encode, Decode)]
+#[cfg_attr(feature = "std", derive(frame_support::DebugNoBound))]
+pub struct Node<T: Config> {
+	voter: Voter<AccountIdOf<T>>,
+	prev: Option<AccountIdOf<T>>,
+	next: Option<AccountIdOf<T>>,
+
+	/// The bag index is not stored in storage, but injected during all fetch operations.
+	#[codec(skip)]
+	pub(crate) bag_upper: VoteWeight,
+}
+
+impl<T: Config> Node<T> {
+	/// Get a node by bag idx and account id.
+	pub fn get(bag_upper: VoteWeight, account_id: &AccountIdOf<T>) -> Option<Node<T>> {
+		crate::VoterNodes::<T>::try_get(account_id).ok().map(|mut node| {
+			node.bag_upper = bag_upper;
+			node
+		})
+	}
+
+	/// Get a node by account id.
+	///
+	/// Note that this must perform two storage lookups: one to identify which bag is appropriate,
+	/// and another to actually fetch the node.
+	pub fn from_id(account_id: &AccountIdOf<T>) -> Option<Node<T>> {
+		let bag = current_bag_for::<T>(account_id)?;
+		Self::get(bag, account_id)
+	}
+
+	/// Get a node by account id, assuming it's in the same bag as this node.
+	pub fn in_bag(&self, account_id: &AccountIdOf<T>) -> Option<Node<T>> {
+		Self::get(self.bag_upper, account_id)
+	}
+
+	/// Put the node back into storage.
+	pub fn put(self) {
+		crate::VoterNodes::<T>::insert(self.voter.id.clone(), self);
+	}
+
+	/// Get the previous node in the bag.
+	pub fn prev(&self) -> Option<Node<T>> {
+		self.prev.as_ref().and_then(|id| self.in_bag(id))
+	}
+
+	/// Get the next node in the bag.
+	pub fn next(&self) -> Option<Node<T>> {
+		self.next.as_ref().and_then(|id| self.in_bag(id))
+	}
+
+	/// Get this voter's voting data.
+	pub fn voting_data(
+		&self,
+		weight_of: impl Fn(&T::AccountId) -> VoteWeight,
+		slashing_spans: &BTreeMap<AccountIdOf<T>, SlashingSpans>,
+	) -> Option<VotingDataOf<T>> {
+		match self.voter.voter_type {
+			VoterType::Validator => Some((
+				self.voter.id.clone(),
+				weight_of(&self.voter.id),
+				sp_std::vec![self.voter.id.clone()],
+			)),
+			VoterType::Nominator => {
+				let Nominations { submitted_in, mut targets, .. } =
+					Nominators::<T>::get(self.voter.id.clone())?;
```

Comment:
Why not store the info in `Nominators` and `Validators` in the respective node? We could then just always use `VoterNodes` to look up the nominator/validator info associated with the accountId

Reply:
Can you elaborate more?

Reply:
@kianenigma If every time we want access a node we are going to do a read of `Nominators::get(voter)` OR `Validators::get(voter)` I think we should consider removing the `Nominators` and `Validators` storagemaps and instead add a field to each Node that is an enum like

```
enum VoterTypeInfo {
   Nominator(Nominations)
   Validator(ValidatorPrefs)
}
```

However, I have not thought about this too much, so I could see it being slower for the average access pattern. For example if you simply wanted to do `Nominators::contains_key(voter)`, you instead would have to read in the Node, decode it, and then check if `VoterTypeInfo` is `VoterTyepInfo::Nominator`.

But I haven't thoughtfully looked at the access patterns and I am wondering if we have considered this

Reply:
your observation is correct. The decision was to make the bags work similar to a component to staking that can live next to it, and be enable/disabled via a future flag or similar. 

For example, I was thinking of even putting it into its own pallet.

Reply:
Closing for now since the plan is to move to its own pallet